Lv2 support2

daschuer opened the pull request:
This PR adds basic LV2 support to Mixxx. It integrates seamlessly with our current Effects Framework. Just install an LV2 plugin and it will show up when you cycle through the effects.

Currently it is only supporting plugins which handle stereo input/output audio samples and do not require any additional features.

This is the successor of 
https://github.com/mixxxdj/mixxx/pull/316

IMHO this is in a mergeable state, because if a user has no LV2 plug-in installed, he will not noticed this branch at all. There are some pending shortcomings when we look at the visual integration and customize the interface of some plug-Ins but that are no stopper for me. 

Lets merge it for now to stop rotting of this already useful branch. If it tuns out that it introduces issue for some regular users, we can opt it out from our release builds. 

What do you think? 

@rryan: What is required to add lilv to our buildservers?

Pegasus-RPG committed:
Add Gemini to USB rules list

christophski commented:
This is very exciting, can't wait to use my CALF reverb plugin live.

rryan commented:
Nice! Looks like a pretty small set of changes on the whole. 

lilv support is waiting for a contributor to pick it up -- there's nothing blocked on me now that buildserver development is automated by pull request:
https://github.com/mixxxdj/buildserver/issues/15

Maybe @sblaisot would be interested in the windows side of that?

Be-ing commented:
This does not build with #1092 merged:
```
src/effects/lv2/lv2manifest.cpp: In constructor ‘LV2Manifest::LV2Manifest(const LilvPlugin*, QHash<QString, LilvNodeImpl*>&)’:
src/effects/lv2/lv2manifest.cpp:68:36: error: ‘SEMANTIC_UNKNOWN’ is not a member of ‘EffectManifestParameter’
             param->setSemanticHint(EffectManifestParameter::SEMANTIC_UNKNOWN);
                                    ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:69:33: error: ‘UNITS_UNKNOWN’ is not a member of ‘EffectManifestParameter’
             param->setUnitsHint(EffectManifestParameter::UNITS_UNKNOWN);
                                 ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:76:39: error: ‘CONTROL_TOGGLE_STEPPING’ is not a member of ‘EffectManifestParameter’
                 param->setControlHint(EffectManifestParameter::CONTROL_TOGGLE_STEPPING);
                                       ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:79:39: error: ‘CONTROL_TOGGLE_STEPPING’ is not a member of ‘EffectManifestParameter’
                 param->setControlHint(EffectManifestParameter::CONTROL_TOGGLE_STEPPING);
                                       ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:81:39: error: ‘CONTROL_KNOB_STEPPING’ is not a member of ‘EffectManifestParameter’
                 param->setControlHint(EffectManifestParameter::CONTROL_KNOB_STEPPING);
                                       ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:83:40: error: ‘CONTROL_KNOB_LINEAR’ is not a member of ‘EffectManifestParameter’
                  param->setControlHint(EffectManifestParameter::CONTROL_KNOB_LINEAR);
                                        ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:108:36: error: ‘SEMANTIC_UNKNOWN’ is not a member of ‘EffectManifestParameter’
             param->setSemanticHint(EffectManifestParameter::SEMANTIC_UNKNOWN);
                                    ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:109:33: error: ‘UNITS_UNKNOWN’ is not a member of ‘EffectManifestParameter’
             param->setUnitsHint(EffectManifestParameter::UNITS_UNKNOWN);
                                 ^~~~~~~~~~~~~~~~~~~~~~~
src/effects/lv2/lv2manifest.cpp:110:35: error: ‘CONTROL_TOGGLE_STEPPING’ is not a member of ‘EffectManifestParameter’
             param->setControlHint(EffectManifestParameter::CONTROL_TOGGLE_STEPPING);
                                   ^~~~~~~~~~~~~~~~~~~~~~~
scons: *** [lin64_build/effects/lv2/lv2manifest.o] Error 1
scons: building terminated because of errors.
```

Be-ing commented:
Thank you for working on this, but I am doubtful it should be rushed for 2.1. Maybe we can hide it behind a compile option that is disabled by default like the OpenGL ES waveform renderer. There's still quite a lot of polishing to do:
* Blacklisting parameters in the preferences does not update loaded effects. Another effect must be loaded then the old one reloaded.
* There are no parameter tooltips. Parameter names are often long, so they need to be elided. Tooltips are important even if they only show the full name of the parameter.
* The big list of buttons for each effect in the preferences is too tall. It should be a list. It should probably also be a part of the main effects preferences instead of its own pane.
* Effects should be able to be blacklisted, not just parameters.
* For a good user experience, we should have default parameter blacklists and metaknob linkings for popular LV2 plugins. There is no way to create those presently.
* Skins need to be able to differentiate between binary switches and buttons with more than 2 states. For example, the Calf Reverb plugin has a "Room Size" parameter which has multiple discrete settings possible, but skins only show ON/OFF. This is confusing.
* At least with #940, for effects with lots of parameters, some knobs do not show until their value is changed.

sblaisot commented:
I will try working on lilv on the windows build server in the next weeks.
btw, how to assign the issue to myself ?

daschuer commented:
Now the parameter slot assignment is refreshed for loaded effects as well. 
This allows assign rarely uses parameters, tweak them and unassigned them. A kind of a "prepare mode".  However, these settings are not persistent as it is most likely expected. 

@Be: What could be a solution? 

We need a way, to store the values of all parameters in the effect.xml file. In addition we need to store the blacklistings. It is also a question what should happen with the hidden parameters. Do we need to store a default, or is the value per effect slot in effect.xml sufficient? 

All this is probably to much to be part of 2.1, but I just want to be sure the effect.xml format is settled.

Be-ing commented:
Hmm, interesting question. I'm not sure. The way I was thinking blacklisting parameters would be used would be to hide parameters whenever that effect is loaded into any effect slot. In that case, effects.xml would not be an appropriate place to store the blacklist. On the other hand, it may be desirable to have some parameters blacklisted when exporting/importing chains.

Be-ing commented on src/preferences/dialog/dlgpreferences.cpp:
```diff
@@ -203,12 +203,11 @@ void DlgPreferences::createIcons() {
+    m_pEffectsButton->setIcon(0, QIcon(":/images/preferences/ic_preferences_effects.png"));
+    m_pEffectsButton->setText(0, tr("Effects"));
+    m_pEffectsButton->setTextAlignment(0, Qt::AlignLeft | Qt::AlignVCenter);
+    m_pEffectsButton->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);
```
There were a few more changes in f6403a21bc2493fde8c5106ddd2216a89cbbb63b to revert.

Commit:
Advanced Mapping for the Behringer CMD MM-1 added

Commit:
forgot to add wiki and forum  url

Commit:
almost everything changed for better fx support XD

Commit:
0 Day patch; leds not updating when in FX-Mode

Commit:
FxUnits now assign themselves to the channel instead of the Channel assigning to the fxUnit

Be-ing commented on src/effects/lv2/lv2effectprocessor.cpp:
```diff
@@ -0,0 +1,102 @@
+    lilv_instance_connect_port(m_handle, audioPortIndices[2], m_outputL);
+    lilv_instance_connect_port(m_handle, audioPortIndices[3], m_outputR);
+
+    lilv_instance_activate(m_handle);
```
This is not the correct place to call this function. The LV2 plugin's `activate()` function needs to be called before `LV2EffectProcessor::process` whenever the effect is enabled. We could hack this into the existing `LV2EffectProcessor::process` function by using the EnableState like native effects, but I think it would be a better idea to add an `activate()` method to the `EffectProcessor` API and refactor the native effects to use it. Then we might be able to get rid of the intermediate ENABLING and DISABLING EffectProcessor::EnableStates.

Be-ing commented on src/effects/lv2/lv2effectprocessor.cpp:
```diff
@@ -0,0 +1,102 @@
+
+    // We assume the audio ports are in the following order:
+    // input_left, input_right, output_left, output_right
+    lilv_instance_connect_port(m_handle, audioPortIndices[0], m_inputL);
```
Can this handle effects with mono inputs and outputs?

Be-ing commented on src/effects/lv2/lv2backend.cpp:
```diff
@@ -0,0 +1,96 @@
+        if (lilv_plugin_is_replaced(plug)) {
+            continue;
+        }
+        LV2Manifest* lv2Manifest = new LV2Manifest(plug, m_properties);
```
This should use lilv_plugin_has_latency to filter out plugins that add latency. We cannot compensate for latency in effects units with switchable input paths because we cannot suddenly switch which channels are being delayed during playback.

daschuer replied:
lilv_plugin_has_latency just tells us if the Plug-In reports latency, not if it introduces it. 
There is nothing wrong to use a Plug-In that introduces Latency if the Mix knob sticks on fully wet. This is also the case for all our fidlib based effects, which introduce ad group delay.

Be-ing replied:
Mixxx can't do anything about plugins that add latency but do not report it. Plugins that add latency are problematic because activating them will create an audible gap in the signal. Even if that is done when the volume fader is down, it would create an odd situation to manipulate different signal paths with different latencies and throw off the alignment of waveforms.

daschuer replied:
It depend on the amount of latency it will add. Currently we fade from the wet to the dry signal by one buffer size. This should cover all known delays. Do you know a plug-In that has a longer latency?

Be-ing commented on src/effects/lv2/lv2effectprocessor.cpp:
```diff
@@ -0,0 +1,102 @@
+                                       QList<int> controlPortIndices) {
+    m_sampleRate = ControlObject::get(ConfigKey("[Master]", "samplerate"));
+ 
+    m_handle = lilv_plugin_instantiate(plugin, m_sampleRate, NULL); 
```
A separate instance of the plugin for each potential input channel is necessary. When multiple inputs are routed to an effect unit with an LV2 plugin loaded, the state is shared between the two channels which produces distorted sounds.

ferranlala committed:
Decrease Reverb Decay

johanLsp committed:
Fix crash when parsing bad xml controller file

johanLsp committed:
Fix deck state when loading empty track

larromba committed:
+ Korg Kaoss DJ XML / JS mappings

larromba committed:
fix: MIDI OUT bug with Mac OS 10.12 & Korg Kaoss DJ 1.6

larromba committed:
fix: beatgrid move bug

larromba committed:
added Fx functionality

When Fx button left is enabled, touchpad can be used to change FX1 rack
mix volume.

When Fx button right is enabled, touchpad can be used to change FX2
rack mix volume.

larromba committed:
Fx rack mix set to 0 on touch up (mimicking serato dj intro)

Commit:
Shift+Filter can control either FX4 or channel gain

Commit:
ShiftFilterFX4 default  value set to true

Be-ing committed:
A&H Xone K2/K1: map loop_in and loop_out with supershift mode

loop_in & loop_out do not require showing any state on the LEDs, so
they work well for supershift mode.

Be-ing committed:
A&H Xone K2/K1: add mode to control all parameters of 1 effect unit

accessed by pressing the Layer button of the 4 effect unit layouts

Be-ing committed:
Components: fix bug with effect selection for some controllers

... that do not have a way to query the controller on startup for the
position of all the knobs.

Be-ing committed:
A&H Xone K2/K1: use shift + knob turn for effect selection

when a single effect unit is focused. Any of the knobs for the effect
parameters can be used.

Be-ing committed:
A&H Xone K2/K1: map master and headphones effect routing buttons

These replace the deck 3 & 4 buttons for 2 deck layouts. For the 4
effect unit layouts, the master and headphone routing buttons are
accessible with shift.

Be-ing committed:
A&H Xone K2/K1: rename files to match conventions

Be-ing committed:
A&H Xone K2/K1: swap beatjumping and loop halve/double functions

In my experience accessing beatjumping quickly is more important
than quickly adjusting the loop size.

esbrandt committed:
Update translation template. Found 2348 source text(s) (7 new and 2341 already existing). Removed 1 obsolete entries

esbrandt committed:
(wix) Update translations from https://www.transifex.com/mixxx-dj-software/mixxxdj-windows-installer/

esbrandt committed:
Pull latest translations from https://www.transifex.com/mixxx-dj-software/mixxxdj/github-release-210/. Compile QM files out of TS files that are used by the localized app

Be-ing committed:
Merge pull request #1559 from Be-ing/hercules_p32_mapping_for_2.1

Hercules P32: add option to toggle effect units with decks

Commit:
Reloop TerminalMix 2/4: adapt fx mapping to rotary encoder

Commit:
Reloop TerminalMix 2/4: implement 2.1 loop controls, implement script.triggerControl

daschuer committed:
Added [Effects] AdoptMetaknobPosition prefeences option

daschuer committed:
added metaknobDefault to effect manifests

Commit:
Reloop TerminalMix 2/4: map BPM tap to Beats encoder press

nieknooijens committed:
add hotcue mappings to hercules DJ RMX on buttons 1-4 and beatloop in/out to 5 and 6

nieknooijens committed:
and remove unhandled button prints

Commit:
Reloop TerminalMix 2/4: remove duplicates, clean up useless controls, add <decription> tags for mixer, FX, looping sections etc.

Commit:
Reloop TerminalMix 2/4: implement [Library],GoToItem

Be-ing committed:
Merge pull request #1490 from ronso0/Reloop-TerminalMix-2/4-update

Reloop TerminalMix 2/4 :: mapping update

dszakallas committed:
mv launchpad novation-launchpad

dszakallas committed:
Update release to 2.1.6

dszakallas committed:
update src to 2.1.6

Be-ing committed:
Merge pull request #1228 from dszakallas/controller/novation-launchpad

Controller mapping: Add mappings for Novation Launchpad and Novation Launchpad MK2

Be-ing committed:
fix controller mapping validation test

working around this error that happens when the Components
EffectUnit is initialized by the test:
ControllerEngine: Watching JS File: "E:/Jenkins/workspace/v2.1/mixxx-2.1-release/architecture/amd64/platform/windows/res/controllers/Reloop Terminal Mix 2-4.js"
ControllerEngine: Loading "E:/Jenkins/workspace/v2.1/mixxx-2.1-release/architecture/amd64/platform/windows/res/controllers/Reloop Terminal Mix 2-4.js"
ControlDoublePrivate::getControl returning NULL for ( "[EffectRack1_EffectUnit1]" , "show_parameters" )
"ControllerEngine: script tried to connect to ControlObject ([EffectRack1_EffectUnit1], show_parameters) which is non-existent, ignoring."
ControllerEngine: "Uncaught exception at line 666 in file E:/Jenkins/workspace/v2.1/mixxx-2.1-release/architecture/amd64/platform/windows/res/controllers/midi-components-0.0.js: TypeError: Result of expression 'this.showParametersConnection' [undefined] is not an object."
ControllerEngine shutting down...
src\test\controller_preset_validation_test.cpp(194): error: Value of: testLoadPreset(preset)
  Actual: false
Expected: true

daschuer committed:
fix merge issue

daschuer committed:
Don't add skin: prefix to an empty path

daschuer committed:
improve PixmapSource::getId() Id generation to avoid collisions that may cause collisions

daschuer committed:
Fix use of uninitalizd Image pointer

daschuer committed:
WImagestore use smartpointers

Be-ing committed:
Tango: revert to raster level meters

Rendering SVG level meters caused the whole GUI to become severely
unresponsive on older Macs:
https://bugs.launchpad.net/mixxx/+bug/1759433

Be-ing committed:
LateNight: revert to raster level meters

Rendering SVGs for the level meters makes the GUI severely laggy
on old Macs:
https://bugs.launchpad.net/mixxx/+bug/1759433

Be-ing committed:
Merge pull request #1517 from Be-ing/restore_metaknob_effect_switch_behavior

reimplement parameters reflecting metaknob when switching effects

Be-ing committed:
clarify function names

Be-ing committed:
Merge pull request #1571 from Be-ing/mapping_validation_test_fix

fix controller mapping validation test

Be-ing committed:
Merge pull request #1572 from daschuer/lp1760170

Try to fix Lp1760170 (Shade crash)

Be-ing committed:
improve description of effect load metaknob behavior options

uklotzde committed:
Update Denon MC6000MK2 mapping

Commit:
changed out knobs to affect balance/gain instead of FX knobs

Commit:
adds function that sets fxunit states to "neutral" (refer to #1103)

Be-ing committed:
Tango: fix missing ".png" extension

Be-ing committed:
Merge pull request #1573 from Be-ing/revert_vectorized_meters

Revert to raster level meters in LateNight and Tango to fix severe GUI lag on macOS

Commit:
final commit after jshint and jsbeautifier checking.

Be-ing committed:
Behringer CMD MM1: remove "Advanced" from mapping name

Be-ing committed:
Behringer CMD MM1: don't turn effects on at startup

I recommended this because I didn't realize that the mapping
has controls for the effect enable buttons.

Be-ing committed:
Deere: fix alignment of PFL buttons in 4 deck mixer when scaled

Be-ing committed:
Deere: fix alignment of crossfader assignment buttons when scaled

Be-ing committed:
Deere: fix scaling of vertical spacing in 4 deck mixer

Commit:
Shade: make Library tree split size persistent

Commit:
LateNight: make EffectSelector smaller so it's not covered by Meta knob. It can expand if the screen allows it

Commit:
LateNight: fix lp:1758552, add blank button icons, remove unused style sheets

Commit:
LateNight: fix inverted Aux play button

Be-ing committed:
Merge pull request #1577 from ronso0/shade-library-splitter-fix

Shade :: make library sidebar split size persistent

Be-ing committed:
Merge pull request #1576 from Be-ing/deere_pfl_button_fix

Deere: fix alignment of PFL buttons in 4 deck mixer when scaled

Be-ing committed:
Deere: redesign fader and knob colors

Be-ing committed:
Merge pull request #1578 from ronso0/LateNight-effect-selector-overlap-fix

LateNight :: fix overlapped effect selector + other small fixes

Be-ing committed:
Merge remote-tracking branch 'upstream/2.1'

Be-ing committed:
Deere: make purple and green knobs a little brighter

Commit:
changed fxmode controls so 3rd level is 1st.

Be-ing committed:
Merge pull request #1580 from Swiftb0y/CMDMM1_final_touches

Switched Controls in FXmode

Be-ing committed:
Deere: remove text labels from Mix and Super knobs

Commit:
Deere: fix cut-off effect selector on Mac

Commit:
Shade: fix cut-off effect selector on Mac

fayaaz committed:
Update targeted version of mixxx

fayaaz committed:
Recording button to update with recording status

fayaaz committed:
Update loop indicator

fayaaz committed:
Update remix sampler buttons to load, unload and stop samples

fayaaz committed:
Increase low value of LEDs and update colours

Be-ing committed:
Deere: fix WEffectSelector text getting cut off on Retina Macs

Be-ing committed:
LateNight: fix WEffectSelector text getting cut off on Retina Macs

Be-ing committed:
Shade: add comment

Be-ing committed:
DlgPrefInterface: show skin preview screenshot

daschuer committed:
Fix shimmering of GL Waveforms

fayaaz committed:
Update mappings contributors

fayaaz committed:
Add on air indicator for recording

fayaaz committed:
Remove commented code

daschuer committed:
Merge pull request #1583 from daschuer/shimmering

Shimmering Fix

fayaaz committed:
remove redundant call in onRecordingChanged

larromba committed:
touchpad x & y axis

larromba committed:
touchpad x & y axis

larromba committed:
const -> var

larromba committed:
const -> var

larromba committed:
cue_cdj -> cue_gotoandplay

larromba committed:
cue_cdj -> cue_gotoandplay

larromba committed:
fix: variable declared as empty function

larromba committed:
fix: variable declared as empty function

Be-ing committed:
Merge pull request #1509 from larromba/feature/korgKaossDJ

Feature/korg kaoss dj

fayaaz committed:
Style fixes

Be-ing committed:
Merge pull request #1584 from fayaaz/2.1-kontrol-s4-mappings-2

Traktor Kontrol S4 Mappings more updates

Be-ing committed:
Merge pull request #1554 from Be-ing/xone_k2

Allen & Heath Xone K2/K1 mapping part 2

Be-ing committed:
DlgPrefInterface: select default skin in combobox with empty profile

Be-ing committed:
Merge pull request #1579 from Be-ing/deere_pfl_button_fix

Deere: redesign fader and knob colors

Be-ing committed:
Kontrol S4: rearrange loop encoder mappings

  * move pitch adjustment to shift + right encoder
  * move pitch reset to shift + right encoder press
  * use beatjump_forward/backward for left encoder turn. This will
    move the loop if one is active
  * use shift + left encoder turn for adjusting beatjump_size
  * use reloop_andstop for shift + left encoder press

plus minor code cleanup

Be-ing committed:
Kontrol S4: keep loop dot position for fractional loop sizes

Be-ing committed:
Deere: fix cut off effect names for Retina Macs... again

daschuer committed:
Merge pull request #1585 from Be-ing/skin_combobox_default

DlgPrefInterface: select default skin in combobox with empty profile

daschuer committed:
use isAdoptMetaknobValueEnabled and "[Effects]", "AdoptMetaknobValue"

Be-ing committed:
Merge pull request #1582 from Be-ing/effectselector_macos_cutoff

Fix effect selector text getting cut off on Retina Macs

daschuer committed:
Don't computeSettingsPath with empty input

esbrandt committed:
Deere: Switch vumeter images from SVG to PNG. This works around lags when rapidly updating svg renderinngs on macOS ( at least with older, intergrated GPUs)

esbrandt committed:
Lower the vumeter <PeakHoldTime> value, which is expensive. Make all vumeter use the same <PeakHoldTime> and <PeakFallTime>

esbrandt committed:
Deere: Fix error in sampler row selection button ``<PushButton>: WPushButton::setup: Connecting a Pushbutton not in PUSH or TRIGGER mode is not implemented`` , that was printed to the log on launch

daschuer committed:
don't try to add hotcue number for non hotcue marks

daschuer committed:
Avoid expected missing control warnings

Be-ing committed:
Merge pull request #1587 from esbrandt/deere-vumeter-fix

[Deere] Performance fix - Switch from SVG to PNG for VuMeters , tweak <PeakHoldTime> value

Be-ing committed:
Merge pull request #18 from daschuer/restore_metaknob_effect_switch_behavior

use isAdoptMetaknobValueEnabled and "[Effects]", "AdoptMetaknobValue"

Be-ing committed:
Merge pull request #1574 from Be-ing/effect_load_metaknob

Effect load metaknob behavior

uklotzde committed:
Merge pull request #1575 from uklotzde/lp1733215_denon-mc6000mk2

lp1733215: Mapping for  DENON MC6000MK2

daschuer committed:
Return QString() to clarify code

daschuer committed:
Improve skin warning in WPushButton

daschuer committed:
Fix skin warnings

daschuer committed:
Fix co counting in DlgPrefVinyl

Be-ing committed:
enable metadata editing in track table by default except for macOS

workaround for https://bugs.launchpad.net/mixxx/+bug/1665583

daschuer committed:
delete all leaking COs

Be-ing committed:
add support for platform-dependent styles

This will allow us to work around platform-dependent quirks
in styling. For 2.1 this is needed to work around the issue of
the checkmark pushing text to the right and getting cut off
in WEffectSelector only on macOS.

daschuer committed:
Fix beat jump and beat loop size boxed

Commit:
Deere: let WEffectSelector shrink, remove obsolete knob color

Commit:
Deere: let effect parameters shrink

Commit:
Deere: reduce minimum window width

Commit:
LateNight: let fx parameters shrink at minimum window idth

Commit:
LateNight: adapt library width to window width

Commit:
Tango: let fx parameters shrink at minimum window width

Be-ing committed:
fix WEffectSelector text getting cut off on macOS

Hopefully for real this time

Be-ing committed:
Shade: fix scaling of WEffectSelector

Be-ing committed:
Shade: re-fix WEffectSelector on macOS

Be-ing committed:
Tango: fix WEffectSelector text cut off on left with KDE

Be-ing committed:
Merge pull request #1592 from ronso0/deere-fix-squeezed-fx-parameters

Deere & LateNight & Tango :: fix lp:1758174 squeezed fx parameters at minimum window width

Be-ing committed:
move #ifdefs to one central location

Be-ing committed:
Merge pull request #1590 from Be-ing/platform_dependent_style

add support for platform-dependent skin styles + fix WEffectSelector styling

Be-ing committed:
Merge pull request #1588 from daschuer/warnings

Remove some warnings from log.

Be-ing committed:
Merge pull request #1568 from Be-ing/tracktable_editing

disable track table editing after clicking selected track by default on macOS

Be-ing committed:
set all skins' minimum height to 500 px

LateNight did not completely fit on my old 1366 x 768 laptop
screen before without full screen mode.

Be-ing committed:
set LateNight and Deere minimum height to 550px

Be-ing committed:
Merge pull request #1593 from Be-ing/skin_height

set LateNight minimum height to 550px

daschuer committed:
fixed SuperToMeta_Softtakeover_EffectEnabled test

Be-ing committed:
add skin preview screenshots

Be-ing committed:
add preview screenshots for "(64 samplers)" skins

Be-ing committed:
show placeholder image if skin does not have a preview screenshot

Be-ing committed:
2.1 RC1 CHANGELOG updates

Be-ing committed:
Merge pull request #1525 from Be-ing/skin_preview

skin preview screenshots in Interface preferences

daschuer committed:
Merge pull request #31 from Be-ing/2.1_rc1

2.1 RC1 CHANGELOG updates

daschuer committed:
Update Translation template. Found 2352 source texts (5 new and 2347 already existing)

daschuer committed:
Try to fix MacOs build

daschuer committed:
Fix interface dlg (again)

daschuer committed:
Merge remote-tracking branch 'upstream/2.1'

daschuer committed:
Merge pull request #1569 from daschuer/2.1.0-rc1

prepare a release candidate for 2.1

Be-ing committed:
do not strip debugging symbols in macOS build

Debugging symbols let us know what is going wrong!

rryan committed:
Merge pull request #1594 from Be-ing/mac_do_not_strip_symbols

do not strip debugging symbols in macOS build

Be-ing committed:
fix effect names getting cut off on Retina Macs in Shade... again

Tested on a 2016 Macbook Pro (2880 x 1800). Increasing the negative
left margin any more makes the text start cutting off on the left
on a Mid 2011 Macbook Air (1440 x 900).

Be-ing committed:
Merge remote-tracking branch 'upstream/2.1'

esbrandt committed:
Update translation template.  Found 2352 source text(s) (5 new and 2347 already existing). Removed 1 obsolete entries. Something what went wrong before, and changes from 97271264f6b8e2de1bd013826c73ce003f33c2cc went missing . This commit adds the lost strings to the template again. Now up-to-date.

esbrandt committed:
Pull latest translations from https://www.transifex.com/mixxx-dj-software/mixxxdj/github-release-210/ . Compile QM files out of TS files that are used by the localized app

esbrandt committed:
Pull latest translations for any new languages over 5% completion that were not previously present from https://www.transifex.com/mixxx-dj-software/mixxxdj/github-release-210/ . Compile QM files out of TS files that are used by the localized app. Add new languages to resource collection file ``res/mixxx.qrc``

daschuer committed:
Added ronso0 to the conributor list. Thank you for all the great work and making Tango real.

Be-ing committed:
Merge remote-tracking branch 'upstream/2.1'

Be-ing committed:
Deere: define text color for WEffectSelect QAbstractItemView

Otherwise the system color is used on Windows instead of inheriting
the text color from WEffectSelector

ywwg committed:
Update config string value for recent mixxx versions.
Delete m_bUpgraded dead code.

ywwg committed:
remove commented out code

Commit:
LateNight: align EQ knobs with Vol fader also without Kill buttons (4decks)

daschuer committed:
Merge pull request #1598 from ronso0/latenight-eq-4decks-fix

LateNight 4 decks :: align EQ knobs with Vol fader also without Kill buttons

ywwg committed:
upgrade string: Also upgrade 2.1rc versions

uklotzde committed:
Fix criticial issues for Denon MC6000MK2

See also: https://www.mixxx.org/wiki/doku.php/denon_mc6000mk2

uklotzde committed:
Merge pull request #1599 from uklotzde/lp1763205_hotfix_denon_mc6000mk2

lp1763205: Hotfix for Denon MC6000MK2

uklotzde committed:
Denon MC6000MK2: Fix navigation & looping

uklotzde committed:
Merge pull request #1600 from uklotzde/denon-mc6000mk2-midi

Denon MC6000MK2: Fix navigation & looping

Tested and documented by myself.

uklotzde committed:
Denon MC6000MK2: Use EffectUnit from components.js

uklotzde committed:
Merge pull request #1601 from uklotzde/denon-mc6000mk2-midi

Denon MC6000MK2: Finish effect unit mapping

ywwg committed:
fix rc string check

ywwg committed:
get the upgrade substring right

Be-ing committed:
fix debug assertion when toggling effect unit enable switch

When changing "[EffectRack1_EffectUnitX], enabled", channels that
had never been enabled were getting the intermediate Disabling
and Enable EffectEnableStates. This resulted in the
"pState != nullptr" debug assertion in
EffectProcessorImpl::process getting triggered because
EffectStates had not been allocated for those channels. I am not
quite sure how, but that sometimes led to a crash when loading
another effect in the chain and toggling the chain's enabled
switch.

Fixing https://bugs.launchpad.net/mixxx/+bug/1748061

daschuer committed:
Merge pull request #1597 from ywwg/upgrade-string

Update config string value for recent mixxx versions.

daschuer committed:
Merge pull request #1602 from Be-ing/effectunit_enable_switch_fix

fix debug assertion and crash when toggling effect unit enable switch

uklotzde committed:
Denon MC6000MK2: Enable/disable effect units with TAP button

uklotzde committed:
Merge pull request #1603 from uklotzde/denon-mc6000mk2-midi

Denon MC6000MK2: Enable/disable effect units with TAP button

Tested and approved by myself ;) Reorderings in XML file are caused by Mixxx when loading it. Too lazy to sort this out.

sohet committed:
VCI-100mkII: replace ChooseItem by GoToItem

Be-ing committed:
Merge pull request #1604 from sohet/2.1

VCI-100mkII: replace ChooseItem by GoToItem

Be-ing committed:
Deere: slightly increase maximum width of effect parameters

6b7c9be772b942c8f0628459b318b69cc93b5703 let parameters shrink
and elide when not enough horizontal space is available. However,
when eliding is turned on, Qt will prematurely elide unless there
are a few extra pixels available. Thus, long effect parameter
names were getting ellipsed even when there was enough horizontal
space available.

Be-ing committed:
LateNight: add option for 4 effect units

Be-ing committed:
Merge pull request #1605 from Be-ing/latenight_4effects

LateNight: add option for 4 effect units

uklotzde committed:
Denon MC6000MK2: Fix components.js usage for cue/play/sync/hotcue

uklotzde committed:
Merge pull request #1606 from uklotzde/denon-mc6000mk2-midi

Denon MC6000MK2: Fix components.js usage

daschuer committed:
Install the PDB files by default

Be-ing committed:
Merge pull request #1607 from daschuer/lp1762737

Wix: Install the PDB files by default

Commit:
LateNight: fix squeezed fx-selector at minimum window width and with Super knobs visible

Commit:
LateNight: relax decks' top row a bit (squeezed Vinyl buttons at minimum window width)

Commit:
LateNight: increase fx-selector width to 102px

Be-ing committed:
Merge pull request #1611 from ronso0/latenight-fx-selector-fix

LateNight :: fix squeezed fx-selector + squeezed Vinyl buttons

nopeppermint committed:
Update deck_row_1_keyVinylFx.xml

nopeppermint committed:
Update vinyl_controls.xml

nopeppermint committed:
Update deck_row_1_keyVinylFx.xml

Be-ing committed:
Merge pull request #1612 from nopeppermint/patch-6

LateNight :: fix more squeezed Vinyl buttons

Be-ing committed:
release Mixxx 2.1 (#1608)

rryan committed:
Update Debian changelog.

daschuer committed:
Merge pull request #1614 from rryan/2.1-changelog

Update Debian changelog.

Be-ing committed:
update manual PDF for 2.1 release (#1615)

uklotzde committed:
Fix invocation args of pasuspender

uklotzde committed:
Swap default values for temp/perm rate changes

daschuer committed:
Merge pull request #1617 from uklotzde/lp1764254_default_rate_change_values

lp1764254: Swap default values for temp/perm rate changes

uklotzde committed:
Fix remaining invocations with of pasuspender

daschuer committed:
Merge tag 'release-2.1.0'

Conflicts:
	src/defs_version.h

daschuer committed:
fix debug assert using empty pixmap source

daschuer committed:
Fix false postive restart request when changing skin.

uklotzde committed:
Fix unresponsive scrolling through crates & playlists with encoder

Be-ing committed:
Merge remote-tracking branch 'upstream/2.1'

daschuer committed:
Merge pull request #1616 from uklotzde/lp1763007_pasuspender_usage

lp1763007: Fix invocation args of pasuspender

daschuer committed:
Merge pull request #1619 from uklotzde/lp1719474_fix_scrolling_through_crates_and_playlists

lp1719474: Fix unresponsive scrolling through crates & playlists using encoder

uklotzde committed:
Merge pull request #1618 from daschuer/211bugfixes

debug assert fix + lp745679 false poisitiv restart request

uklotzde committed:
Use a default timeout of 250 ms between selection and activation

rryan committed:
Update my name in the credits.

rryan committed:
Version bump to 2.1.1-pre.

rryan committed:
Update README and manual link to 2.2.

uklotzde committed:
Distinguish between mouse and keyboard/encoder events in sidebar tree

Be-ing commented:
I would like to have this merged eventually, but there are currently lots of merge conflicts, usability issues, and fundamental problems with how the DSP is handled so I am closing it for now. Whenever work on it resumes, please reopen this PR.

Be-ing committed:
Merge pull request #1621 from rryan/2.1

Version bump to 2.1.1-pre.

Be-ing committed:
Merge remote-tracking branch 'upstream/2.1'

daschuer commented:
I am still maintaining this.

uklotzde committed:
Fork Fidlib

uklotzde committed:
Thread-safe invocation of code for generating filters

uklotzde committed:
Fix compiler warnings

nopeppermint committed:
Tango: change extras to misc

nopeppermint committed:
Tango64: change Extra to Misc

daschuer committed:
Merge pull request #1625 from uklotzde/lp1765210_fidlib

lp1765210 Fidlib: Thread-safe and reentrant generation of filters

daschuer committed:
Merge pull request #1620 from uklotzde/lp1719474_followup

lp1719474 Follow-up: Fix navigation usability issues in sidebar tree

rryan committed:
Workaround for travis-ci/travis-ci#6307.

rryan committed:
Workaround for travis-ci/travis-ci#6307.

daschuer committed:
Merge pull request #1629 from nopeppermint/patch-6

Tango: change name of skin menu "Extras" to "Misc"

nopeppermint committed:
Latenight: group FX Buttons

daschuer committed:
Merge pull request #1634 from nopeppermint/patch-6

Latenight fix: group FX Buttons in deck

daschuer committed:
Merge remote-tracking branch 'upstream/2.1' into master

daschuer committed:
Merge remote-tracking branch 'upstream/master' into lv2_support2

Conflicts:
	.travis.yml
	src/effects/effect.cpp
	src/effects/effect.h
	src/effects/effectchainmanager.h
	src/effects/effectinstantiator.h
	src/effects/effectmanifest.h
	src/effects/effectrack.cpp
	src/effects/native/autopaneffect.cpp
	src/effects/native/balanceeffect.cpp
	src/effects/native/biquadfullkilleqeffect.h
	src/effects/native/bitcrushereffect.cpp
	src/effects/native/echoeffect.cpp
	src/effects/native/echoeffect.h
	src/effects/native/filtereffect.cpp
	src/effects/native/flangereffect.cpp
	src/effects/native/loudnesscontoureffect.cpp
	src/effects/native/phasereffect.cpp
	src/effects/native/reverbeffect.cpp
	src/effects/native/threebandbiquadeqeffect.h
	src/engine/effects/engineeffect.cpp
	src/engine/effects/engineeffect.h
	src/mixxx.cpp
	src/preferences/dialog/dlgpreferences.cpp
	src/preferences/dialog/dlgpreferences.h
	src/test/metaknob_link_test.cpp
	src/test/nativeeffects_test.cpp

daschuer commented:
This one is head up with master again. It was a lot of work to put it back to this state. It is odd to let it rod in a PR again. It works well, but has issues in presenting the LV2 effects in the GUI. 
We may consider to merge this to master and disable it via command line option.

Be-ing reviewed (changes requested):
Thanks for resolving the numerous merge conflicts. I have looked briefly at the code but not done a very thorough review yet. There are still many UX quirks that make this not ready to expose to users yet:
  * There needs to be a way to hide effects. Without this, the list of effects to select can easily grow overwhelmingly big. Also, some detected LV2 effects may be useless. For example, I loaded Gx-DelayStereo and it just made a horrible buzzing noise.
  * There is no way to see plugins' GUIs. Metering and other analysis plugins are useless without being able to see the plugin's GUI. For some plugins, it is not clear how to use their parameters without seeing the plugin's GUI.
  * Our skins currently only support binary effect button parameters but LV2 parameters can have more than 2 states. I am not sure these should be button parameters. It may work better to create a new parameter type that is shown with a new QComboBox subclass.
  * Blacklisted parameters are coupled for every instance of an effect. This would be awkward with saving & loading chain presets. Instead, hidden parameters and parameter order should be independent for every instance of an effect. The default parameters shown for an effect and their order should be stored as part of the user's stored default snapshot for the effect along with parameter values and metaknob linkings. This does not need to be done for the initial merge of this branch. @kshitij98 could refactor how parameter hiding is implemented as part of his GSOC project.
  * The LV2 and Effects preference pages are redundant.

I think we could merge this for 2.2 if you implement at least blacklisting effects.

daschuer commented on src/effects/effectprocessor.h:
```diff
@@ -44,15 +45,6 @@ class EffectState {
-
-    // TODO: implement this for all subclasses and call it when the buffer
-    // size and sample rate are configured by SoundManager
-    virtual void engineParametersChanged(const mixxx::EngineParameters& bufferParameters) {
```
It was unused.
We cannot change the buffer parameter on the fly all over the code. And I am in doubt if we ever need this.
If we, we can implement it with a fresh eye on this.

daschuer replied:
It is not a good practice to keep unused code just as a reminder for a TODO.
I have just filed a bug https://bugs.launchpad.net/mixxx/+bug/1770186

daschuer committed:
fix remaining conflict

Be-ing commented:
@rryan are the build servers set up with lilv?

Be-ing commented on src/effects/lv2/lv2effectprocessor.cpp:
```diff
@@ -0,0 +1,246 @@
+#include "control/controlobject.h"
+#include "util/sample.h"
+
+#define MAX_BUFFER_LEN 160000
```
Replace this with `#include "util/defs.h"`

Be-ing commented on src/mixxx.cpp:
```diff
@@ -197,6 +200,12 @@ void MixxxMainWindow::initialize(QApplication* pApp, const CmdlineArgs& args) {
+    LV2Backend* pLV2Backend = new LV2Backend(m_pEffectsManager);
+    m_pEffectsManager->addEffectsBackend(pLV2Backend);
+#else
+    LV2Backend* pLV2Backend = 0;
```
`0`? Should this be `nullptr`?

Be-ing commented on src/mixxx.cpp:
```diff
@@ -357,7 +366,7 @@ void MixxxMainWindow::initialize(QApplication* pApp, const CmdlineArgs& args) {
     // Initialize preference dialog
     m_pPrefDlg = new DlgPreferences(this, m_pSkinLoader, m_pSoundManager, m_pPlayerManager,
-                                    m_pControllerManager, m_pVCManager, m_pEffectsManager,
+                                    m_pControllerManager, m_pVCManager,  pLV2Backend, m_pEffectsManager,
```
extra space after `m_pVCManager, `

rryan commented:
Nope: https://github.com/mixxxdj/buildserver/issues/15



On Sat, May 12, 2018, 9:15 AM Be <notifications@github.com> wrote:

> @rryan <https://github.com/rryan> are the build servers set up with lilv?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1240#issuecomment-388565660>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABnn5viU-O7kmYb2fWHR8S8AGz7v30vks5txwqXgaJpZM4M_8Rr>
> .
>

daschuer committed:
Replace tabs by spaces

daschuer committed:
Include "util/defs.h" for MAX_BUFFER_LEN

daschuer committed:
use nullptr instead of plain 0

daschuer committed:
remove extra space

Be-ing commented:
`liblilv-dev` needs to be added to the Build-Depends package list in `build/debian/control`.

Be-ing commented:
Compilation warning:
```
g++ -o lin64_build/engine/enginebuffer.o -c -std=c++11 -pipe -Wall -Wextra -g -fPIC -O3 -ffast-math -funroll-loops -fomit-frame-pointer -mtune=generic -pthread -Dx86_64 -DMIXXX_BUILD_DEBUG -D__LINUX__ -D__UNIX__ -DSETTINGS_PATH=\".mixxx/\" -DSETTINGS_FILE=\"mixxx.cfg\" -DUNIX_SHARE_PATH=\"/home/be/local/share/mixxx\" -DUNIX_LIB_PATH=\"/home/be/local/lib/mixxx\" -D__PORTAUDIO__ -DQT_TABLET_SUPPORT -DQT_SHARED -DQT_DISABLE_DEPRECATED_BEFORE -DQT_CORE_LIB -DQT_GUI_LIB -DQT_OPENGL_LIB -DQT_XML_LIB -DQT_SVG_LIB -DQT_SQL_LIB -DQT_SCRIPT_LIB -DQT_NETWORK_LIB -DQT_WIDGETS_LIB -DQT_OPENGL_LIB -DQT_XML_LIB -DQT_SVG_LIB -DQT_SQL_LIB -DQT_SCRIPT_LIB -DQT_NETWORK_LIB -DQT_TESTLIB_LIB -DQT_SCRIPTTOOLS_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CONCURRENT_LIB -DQT_DBUS_LIB -DQT_CORE_LIB -D__SNDFILE__ -DSFC_SUPPORTS_SET_COMPRESSION_LEVEL -D__MAD__ -D__HID__ -D__BULK__ -D__VINYLCONTROL__ -D__BROADCAST__ -D__VAMP__ -Dkiss_fft_scalar=double -D__SQLITE3__ -D__BATTERY__ -Ilin64_build -Isrc -Ilib/soundtouch-2.0.0 -Ilib/replaygain -Ilib/libebur128/ebur128 -I/usr/include/alsa -I/usr/include/qt5/Qt5DBus -I/usr/include/qt5/QtCore -I/usr/include/qt5 -I/usr/include/qt5/QtGui -I/usr/include/qt5/QtOpenGL -I/usr/include/qt5/QtWidgets -I/usr/include/qt5/QtXml -I/usr/include/qt5/QtSvg -I/usr/include/qt5/QtSql -I/usr/include/qt5/QtScript -I/usr/include/qt5/QtNetwork -I/usr/include/qt5/QtTest -I/usr/include/qt5/QtScriptTools -I/usr/include/qt5/QtConcurrent -I/usr/include/qt5/QtDBus -Ilib/gtest-1.7.0/include -Ilib/fidlib -I/usr/include/taglib -Ilib/qtscript-bytearray -Ilib/reverb -Ilib/portaudio -I/usr/include/libusb-1.0 -Ilib/hidapi-0.8.0-rc1/hidapi -Ilib/xwax -Ilib/scratchlib -Ilib/vamp -I/usr/include/libupower-glib -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include src/engine/enginebuffer.cpp
In file included from /usr/include/qt5/QtCore/qmetatype.h:47,
                 from /usr/include/qt5/QtCore/qobject.h:54,
                 from /usr/include/qt5/QtCore/qiodevice.h:45,
                 from /usr/include/qt5/QtCore/qtextstream.h:43,
                 from /usr/include/qt5/QtCore/qdebug.h:49,
                 from /usr/include/qt5/QtCore/QtDebug:1,
                 from src/util/assert.h:3,
                 from src/util/math.h:23,
                 from src/util/types.h:7,
                 from src/engine/effects/engineeffectchain.h:9,
                 from src/engine/effects/engineeffectchain.cpp:1:
/usr/include/qt5/QtCore/qvarlengtharray.h: In instantiation of ‘void QVarLengthArray<T, Prealloc>::realloc(int, int) [with T = ChannelHandleMap<EngineEffectChain::ChannelStatus>; int Prealloc = 256]’:
/usr/include/qt5/QtCore/qvarlengtharray.h:276:3:   required from ‘void QVarLengthArray<T, Prealloc>::resize(int) [with T = ChannelHandleMap<EngineEffectChain::ChannelStatus>; int Prealloc = 256]’
src/engine/channelhandle.h:209:13:   required from ‘void ChannelHandleMap<T>::maybeExpand(int) [with T = ChannelHandleMap<EngineEffectChain::ChannelStatus>]’
src/engine/channelhandle.h:173:9:   required from ‘void ChannelHandleMap<T>::insert(const ChannelHandle&, const T&) [with T = ChannelHandleMap<EngineEffectChain::ChannelStatus>]’
src/engine/effects/engineeffectchain.cpp:24:85:   required from here
/usr/include/qt5/QtCore/qvarlengtharray.h:392:19: warning: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘class ChannelHandleMap<EngineEffectChain::ChannelStatus>’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
             memcpy(ptr, oldPtr, copySize * sizeof(T));
             ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/engine/effects/engineeffectchain.h:12,
                 from src/engine/effects/engineeffectchain.cpp:1:
src/engine/channelhandle.h:153:7: note: ‘class ChannelHandleMap<EngineEffectChain::ChannelStatus>’ declared here
 class ChannelHandleMap {
       ^~~~~~~~~~~~~~~~
```

Be-ing commented:
I am merging this to the upstream lv2_support branch and will be following up with another PR targeted at that branch.
